src/core/hash.rs

```diff
@@ -19,30 +23,123 @@ pub trait Hash: Debug + Clone + 'static {
         + PartialOrd;
 
     fn hash(s: &[u8]) -> Self::Output;
+
+    fn update(&mut self, s: &[u8]);
+
+    fn finalize(self) -> Self::Output;
+}
+
+#[derive(Debug, Clone)]
+pub struct Sha3Hasher(Option<CoreWrapper<Sha3_256Core>>);
+
+impl serde::ser::Serialize for Sha3Hasher {
```

Comment:
I don't like this. I'm not sure how can we get rid of this Serialize, Deserialize requirement.

Reply:
That is because both `SignatureScheme` and `Hash` are used as generic in `Block` and then `Block` use bincode to serialize/deserialize. Also the `serde` need all the generics with bound `serializer`/`deserialize`. The `SignatureScheme` is bound up with this.